### ISSUE TITLE
Detected MIME/format changed from 0.9 to 1.0

### ISSUE THREAD

**Issue by vladra** [hanami/controller#225]
During migration to Hanami 1.0 (from 0.9), I've noticed that now my `json` actions require to specify `self.format = :json` or response will be returned as `html` otherwise.

Accordingly to gitter chat with @jodosha , it seems that bug could be introduced in #206 or #191

**Comment by jodosha** [comment]
@vladra Thanks for reporting this. It would be really helpful if you can report a dump of a Rack env instance that used to work. Thank you very much.

**Comment by mereghost** [comment]
More info from gitter: the `Accept` header was `application/json,text/plain,*/*`

**Comment by vladra** [comment]
I've tried to write test that fails for hanami/controller, but wasn't able to unfortunately.

Problem I have is coming from `Hanami::Action::Mime#best_q_match`

Incoming header is `"application/json, text/plain, */*"`, available mimes is `["application/json"]`

Then values are determined as `[["application/json", 1.0], ["text/html", 1.0]]`. Sorting keeps order same, and taking last value results in `text/html` response format.

I've added failing test to this commit so you can check when it fails. Test fails only if `default_request_format` is not set (by default in Hanami app it is set to `:html`)

**Comment by vladra** [comment]
Btw using not patched rack method best_q_match works fine, because it is picking first match (in hanami it is last match)

**Comment by mereghost** [comment]
Heya @vladra , thanks for investigating the issue. I took a quick look and using rack's implementation of `best_q_matches` breaks a couple of tests on `controller`.

By the HTTP spec when no `q` value is passed the server can return any of the formats requested and because the `Accept`header said it can return whatever (`*/*`) so `controller` is returning `text/plain` as it's its current default format (rack's seems to be `text/html`).

Not sure on how to tackle this one. =/

**Comment by vladra** [comment]
@mereghost sorry, I forgot to provide link to failing test example 🙂 

https://github.com/vladra/controller/commit/ee3a724a5f4451323b82b1369be550b7841e3a83

well, available mimes are set to `application/json`, so I'm not sure that it is correct behaviour to return `text/html` in response. And, as I said, Racks `best_q_mathces` method returns correct result, because it takes first value from array of mime types. Hanami takes last one. This array looks like this - `[["application/json", 1.0], ["text/html", 1.0]]`.

**Comment by vladra** [comment]
Guys, any suggestions on this?

I could invest some time into this, but not sure in what direction to move.

To conclude:

- bug happens only if `default_response_format` is set. Check failing test example
- original Rack `best_q_match` method returns correct result to me (first from array)
- when I specify `accept :json`, it seems `default_response_format` adds also html to the array of possible mime types. 

So, based on this info, `best_q_match` generated allowed mime types array looks wrong to me. I have `accept :json`, explicitly saying the only format Action should receive.

Thanks for looking here 🙂

**Comment by mereghost** [comment]
That the matches looks kinda weird is true. I've been checking this for the past day and there's no easy answer as per the [Accept Header Spec](https://www.w3.org/Protocols/rfc2616/rfc2616-sec14.html) without a `q` value any of the higher available types can be returned.

Also, rack's `best_q_matches` breaks controller specially on jruby it seems. I'll do some more tests and report any findings here.

**Comment by vladra** [comment]
I investigated it a bit and it looks like PR #170 to support old IE is breaking this. Can you check please?

Without those line to support IE<=8, default response wont be added to the array of allowed mime types. Check #167 comments.

Cause the main problem in my opinion is that Hanami allows response to be html format, despite me forbidding this (thats the role of `accept :json`, right?)

If you agree on this, maybe that PR should be adjusted to fix only IE8 case, not touching other cases. WDYT?

**Comment by vladra** [comment]
I've tried to fix this bug, limiting IE8 case by user agent (case described above), so it is not globally applied for all cases

My test is passes now, but [this test fails then](https://github.com/hanami/controller/blob/master/test/action/format_test.rb#L64) 😕 

It's logical to me though. It wan't failing only because `default_response_format` was not set. Setting `default_response_format` to any value except `html` will also fail that test.

If you interested, please [check this code](https://github.com/vladra/controller/pull/1)

It seems I'm in the dead end right now. Something is wrong there I think😄

**Comment by mereghost** [comment]
@vladra Tanks for investigating further. I'll take a look at your findings and keep you posted.

**Comment by vladra** [comment]
@mereghost thanks, looking forward

one more thing to say so you have full information.

I've just added IE8 specific test (based on this headers [1](https://developer.mozilla.org/en-US/docs/Web/HTTP/Content_negotiation/List_of_default_Accept_values)). And guess what, it is failing with current `best_q_match` implementation (and with my code).

Adding  `default_request_format :html` will make test work though. This is the only reason why case described in #170 is working. Cause everyone who is using Hanami has `default_request_format :html` by default.

**Comment by mereghost** [comment]
@vladra Quick clarification: did your action had an `accept :json`? The test you created had one, but I'm trying to cover all bases here.

**Comment by vladra** [comment]
@mereghost yes, action has `accept :json` inside. This action returns `json` only. Let me know if you need any help

**Comment by dikond** [comment]
Guys, does [this guide](http://hanamirb.org/guides/actions/mime-types/) relevant? I've read **Whitelisting** paragraph and expected that if in my action it will be
```ruby
accept :html
```
and my request headers will have
```
Accept:application/json, text/plain, */*
X-Requested-With:XMLHttpRequest
```
the server would return `406 Not Acceptable` status, but it returns `status 200` with 
```
Content-Type:text/html; charset=utf-8
```

Is this intentional behavior? As far as I understand, Chrome automatically adds `*/*` and we get a response in default format which is `text/html` by default.

**Comment by mereghost** [comment]
@dikond If you have a `*/*` on your accept header, you're basically saying that you accept anything the server wishes to return. If you remove the media range, then it should return a `406`.

**Comment by mereghost** [comment]
@vladra I'm working on a fix for this, but it turned to be way more convoluted than just picking first or last. Also, it seems that Rack 2.0 went with our `last` call instead of with first.

**Comment by vladra** [comment]
@mereghost thanks, let me know if you need my help

**Comment by jodosha** [comment]
Fixed by #229
